### PR TITLE
feat(pt_expt): support .pt training checkpoints in DeepEval

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
## Summary

- `dp --pt-expt test -m foo.pt` previously rejected `.pt` files (only `.pt2` / `.pte` were supported); `dp --pt test -m foo.pt` on a pt_expt-trained checkpoint silently loaded random weights because the dpmodel `.w`/`.b` naming doesn't match the legacy pt backend's `.matrix`/`.bias`.
- This PR makes `.pt` training checkpoints first-class for inference under the pt_expt backend.

## Changes

- **`Backend.detect_backend_by_model`** sniffs `.pt` content and routes by parameter naming: `.w`/`.b` → pt-expt, `.matrix`/`.bias` → pt. Bogus `.pt` falls back to suffix dispatch (pt). Backwards compatible with all existing pt-trained `.pt` checkpoints.
- **`pt_expt.DeepEval._load_pt`** reconstructs the model from `_extra_state[\"model_params\"]`, loads the state-dict via `ModelWrapper`, and exposes an eager `forward_common_lower` runner with the same signature as the AOTI/exported module so the existing `eval()` path is unchanged. Spin-aware (7-arg) and non-spin (6-arg) variants. Multi-task `.pt` selects a head and remaps keys. Populates `metadata` (default_fparam, dim_fparam/aparam, …) so eval helpers behave the same as the `.pt2`/`.pte` path.
- **`pt_expt.get_model`** learns `get_spin_model` (mirrors dpmodel) so spin checkpoints can be reconstructed from `model_params` (previously it silently returned a non-spin `EnergyModel`).
- **Dispatch**: pt_expt's DeepEval ctor now explicitly accepts `.pt2`/`.pte`/`.pt` and raises an actionable `ValueError` for anything else (was: implicit fallthrough to `.pte` loader → cryptic torch error).

## Tests

`source/tests/pt_expt/infer/test_deep_eval_pt_checkpoint.py` (21 tests):

- **Dispatch sniffing** — pt_expt-style `.pt` routes to pt-expt; pt-style `.pt` routes to pt; bogus `.pt` falls back to suffix.
- **Single-task `.pt`** — metadata accessors, `DeepPot(.pt).eval(...)` parity vs direct forward at 1e-10, `.pth` rejection.
- **Multi-task `.pt`** — head selection parity, missing-head error, no-default-no-head error.
- **Spin `.pt`** — metadata flags, eager-reference parity, missing-spin-arg error.
- **Spin + fparam `.pt`** — default fparam matches explicit; varying fparam changes output.
- **Spin + aparam `.pt`** — aparam takes effect; missing-aparam raises.
- **Spin multi-task `.pt`** — each head matches its own eager reference; distinct heads produce distinct outputs.
- **Cross-format `.pt` ↔ `.pte` consistency at 1e-10** for vanilla spin (atomic=True), default fparam (atomic=True), and aparam (atomic=True).

## Test plan

- [ ] `pytest source/tests/pt_expt/infer/test_deep_eval_pt_checkpoint.py -v`
- [ ] `pytest source/tests/pt_expt/infer/ -v` (regression: existing `.pt2`/`.pte` paths)
- [ ] Train a small example with `dp --pt-expt train`, then `dp --pt-expt test -m model.ckpt-N.pt` produces identical metrics to `dp --pt-expt test -m frozen.pt2`

## Known limitations

- pt_expt training itself still has no multi-task or multi-task-spin path; `_load_pt` handles such checkpoints, but a user can't currently produce one via `dp --pt-expt train`. Tests construct them synthetically.
- `_load_pt`'s `exported_module` is a Python closure (eager), not a real `torch.nn.Module`. Sufficient for `dp test`, but `eval_descriptor` / `eval_typeebd` / `eval_fitting_last_layer` won't work from a `.pt` (only from `.pt2`/`.pte`).
- NoPBC `.pt` ↔ `.pte` consistency not separately asserted (same eager code path as PBC).

## Summary by CodeRabbit

* **New Features**
  * Load .pt training checkpoints and select multi-task model heads at load time.
  * New spin-energy model construction and inference with distinct spin vs non-spin exported behavior.

* **Bug Fixes**
  * Improved .pt backend detection with safer inspection and clearer errors for unsupported extensions.

* **Tests**
  * End-to-end tests for checkpoint detection, multi-head routing, eager inference consistency, and spin-model agreement.

### CODE REVIEW

**Review by anyangml — APPROVED** [review]
LGTM

File: deepmd/backend/backend.py [hunk 1]
```diff
@@ -101,10 +101,33 @@ def detect_backend_by_model(filename: str) -> type["Backend"]:
...
-        filename = str(filename).lower()
+        filename_lower = str(filename).lower()
+        # `.pt` is shared between the pt and pt_expt backends. They use
+        # different parameter naming (pt: `.matrix`/`.bias`, pt_expt:
+        # `.w`/`.b`), so peek at the state-dict keys to disambiguate.
+        if filename_lower.endswith(".pt"):
+            try:
+                import torch
+
+                sd = torch.load(filename, map_location="cpu", weights_only=False)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/backend/backend.py]: Valid concern. Fixed in b3bab4bb — both sites now use `weights_only=True`, matching the training resume path at training.py:712. Verified the change still loads real compiled multi-task checkpoints (`_extra_state["model_params"]` deserialises fine under safe mode) and the 23 unit tests still pass.

File: deepmd/pt_expt/infer/deep_eval.py [hunk 2]
```diff
@@ -206,6 +214,160 @@ def _load_pt2(self, model_file: str) -> None:
...
+        )
+
+        from deepmd.pt.utils.env import (
+            DEVICE,
+        )
+        from deepmd.pt_expt.model import (
+            get_model,
+        )
+
+        state_dict = torch.load(model_file, map_location=DEVICE, weights_only=False)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/infer/deep_eval.py]: Valid. Fixed in b3bab4bb — `_load_pt` now uses `weights_only=True`, matching training.py:712. The state-dict's `_extra_state["model_params"]` is plain dict/list/str/number and deserialises under safe mode; verified end-to-end against a real compiled multi-task checkpoint.

File: deepmd/backend/backend.py [hunk 3]
```diff
@@ -101,10 +101,33 @@ def detect_backend_by_model(filename: str) -> type["Backend"]:
...
+                if has_pt_expt and not has_pt:
+                    target_name = "pt-expt"
+                else:
+                    target_name = "pt"
+                for key, backend in Backend.get_backends().items():
+                    if key == target_name:
+                        return backend
+            except Exception:
+                # Fall through to suffix matching if sniffing fails.
+                pass
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/backend/backend.py]: Same finding as the codex P1 above. Already fixed in b3bab4bb — both `Backend.detect_backend_by_model` and `pt_expt.DeepEval._load_pt` now use `weights_only=True`, matching training.py:712.

File: source/tests/pt_expt/infer/test_deep_eval_pt_checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,790 @@
...
+                err_msg=f"head={head}, energy",
+            )
+            np.testing.assert_allclose(
+                f,
+                ref["force"].detach().cpu().numpy(),
+                rtol=1e-10,
+                atol=1e-10,
+                err_msg=f"head={head}, force",
+            )
+            self.assertEqual(de.get_type_map(), src.get_type_map())
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/infer/test_deep_eval_pt_checkpoint.py:463]: Valid. Fixed in d3a57f25 — `_build_model_and_params` now takes a `seed` arg, `TestPtExptLoadPtMultiTask` builds head_a with `seed=42` and head_b with `seed=7`, and a new `test_distinct_heads_produce_distinct_outputs` asserts the two heads actually produce different energies (mirrors the spin variant). Verified: head_a/head_b produce different outputs as expected.

File: source/tests/pt_expt/infer/test_deep_eval_pt_checkpoint.py [hunk 5]
```diff
@@ -0,0 +1,968 @@
...
+        natoms = 4
+        coords = rng.random((1, natoms, 3)) * 8.0
+        cells = np.eye(3).reshape(1, 9) * 10.0
+        atom_types = np.array([i % 2 for i in range(natoms)], dtype=np.int32)
+
+        for head, src in (("head_a", self.model_a), ("head_b", self.model_b)):
+            # Build a DeepPot wrapping this DeepEval for end-to-end eval.
+            dp = DeepPot(self.pt_path, head=head)
+            de = dp.deep_eval
+            e, f, v = dp.eval(coords, cells, atom_types, atomic=False)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/infer/test_deep_eval_pt_checkpoint.py]: Valid. Fixed in d3a57f25 — prefixed all 7 unused unpack variables (across 4 unpack sites at lines 426, 472, 613, 918) with `_`. `ruff check --select=RUF059` is now clean. For the spin eager-vs-eager test, asserting `av` (atom_virial) at this layer is left for a follow-up; it's eager-vs-eager so should match, but it interacts with the `.pt` vs `.pte` spin atom_virial divergence already documented in this PR — better isolated.

File: deepmd/backend/backend.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/backend/backend.py]: Same finding as the codex P1 above. Already fixed in b3bab4bb — both `Backend.detect_backend_by_model` (now reorganised into per-backend `match_filename` in bd3af6ed) and `pt_expt.DeepEval._load_pt` use `weights_only=True`.

File: deepmd/pt_expt/infer/deep_eval.py [hunk 6]
```diff
@@ -206,6 +214,178 @@ def _load_pt2(self, model_file: str) -> None:
         self._pt2_runner = aoti_load_package(model_file)
         self.exported_module = None
 
+    def _load_pt(self, model_file: str, head: str | None = None) -> None:
+        """Load a `.pt` training checkpoint (eager mode, no torch.export)."""
+        from copy import (
+            deepcopy,
+        )
+
+        from deepmd.pt.utils.env import (
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/infer/deep_eval.py]: Valid. Fixed in fa2387b7 — now imports `DEVICE` from `deepmd.pt_expt.utils.env`.

File: deepmd/pt_expt/infer/deep_eval.py [hunk 7]
```diff
@@ -206,6 +214,178 @@ def _load_pt2(self, model_file: str) -> None:
...
+                )
+            head_params = model_params["model_dict"][head]
+            # Restrict state_dict to the chosen head and rename to "Default".
+            head_state = {"_extra_state": state_dict["_extra_state"]}
+            for key, value in state_dict.items():
+                prefix = f"model.{head}."
+                if key.startswith(prefix):
+                    head_state[key.replace(prefix, "model.Default.")] = (
+                        value.clone() if torch.is_tensor(value) else value
+                    )
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/infer/deep_eval.py]: Valid. Fixed in fa2387b7 — `load_state_dict` does not mutate the input dict, so the `.clone()` was pure overhead on large multi-task checkpoints.

File: deepmd/pt_expt/infer/deep_eval.py [hunk 8]
```diff
@@ -206,6 +214,178 @@ def _load_pt2(self, model_file: str) -> None:
...
+        )
+        from deepmd.pt_expt.model import (
+            get_model,
+        )
+
+        state_dict = torch.load(model_file, map_location=DEVICE, weights_only=False)
+        if "model" in state_dict:
+            state_dict = state_dict["model"]
+        model_params = deepcopy(state_dict["_extra_state"]["model_params"])
+
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/infer/deep_eval.py]: Valid. Fixed in fa2387b7 — added an explicit guard that raises `ValueError` naming the expected structure and pointing the user at `dp --pt` / `.pte` / `.pt2` alternatives.

File: deepmd/backend/backend.py [hunk 9]
```diff
@@ -101,10 +101,33 @@ def detect_backend_by_model(filename: str) -> type["Backend"]:
...
+                keys = list(sd.keys()) if hasattr(sd, "keys") else []
+                has_pt_expt = any(k.endswith(".w") or k.endswith(".b") for k in keys)
+                has_pt = any(k.endswith(".matrix") or k.endswith(".bias") for k in keys)
+                if has_pt_expt and not has_pt:
+                    target_name = "pt-expt"
+                else:
+                    target_name = "pt"
+                for key, backend in Backend.get_backends().items():
+                    if key == target_name:
+                        return backend
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/backend/backend.py]: Obsolete after the refactor in bd3af6ed: that whole `target_name` lookup loop was removed — `Backend.detect_backend_by_model` now uses a generic specificity score (`Backend.match_filename`), and per-backend overrides (`PyTorchExportableBackend.match_filename`) own their own claim logic. No more manual lookup needed.

File: source/tests/pt_expt/infer/test_deep_eval_pt_checkpoint.py [hunk 10]
```diff
@@ -0,0 +1,968 @@
...
+        cls.files = _make_spin_files(cls.spin_config)
+        cls.model = cls.files["model"]
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        for ext in (".pt", ".pte"):
+            path = cls.files[ext]
+            if os.path.exists(path):
+                os.unlink(path)
+        os.rmdir(cls.files["tmpdir"])
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/infer/test_deep_eval_pt_checkpoint.py]: Valid. Fixed in fa2387b7 — `shutil.rmtree(tmpdir, ignore_errors=True)` replaces the per-file unlink + `os.rmdir`.

File: deepmd/backend/pt_expt.py [hunk 11]
```diff
@@ -44,6 +44,43 @@
...
+            # untrusted .pt — sniffing only needs the dict keys.
+            sd = torch.load(filename, map_location="cpu", weights_only=True)
+            if isinstance(sd, dict) and "model" in sd:
+                sd = sd["model"]
+            keys = list(sd.keys()) if hasattr(sd, "keys") else []
+            has_pt_expt = any(k.endswith(".w") or k.endswith(".b") for k in keys)
+            has_pt = any(k.endswith(".matrix") or k.endswith(".bias") for k in keys)
+            if has_pt_expt and not has_pt:
+                return 2
+        except Exception:
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/backend/pt_expt.py:73]: Valid (CodeQL "Empty except"). Fixed in 042e16b9 — narrowed the try-block to wrap only `torch.load`, added an explanatory comment for why a load failure surrenders the backend claim (lets the dispatcher fall back to pt's default suffix match), and replaced `pass` with an explicit `return 0`.

File: deepmd/backend/backend.py [hunk 12]
```diff
@@ -101,10 +101,35 @@ def detect_backend_by_model(filename: str) -> type["Backend"]:
...
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/backend/backend.py]: Already obsolete. The hard-coded `target_name` block this comment refers to (in commit d3a57f25) was completely removed in the refactor commit bd3af6ed — `Backend.detect_backend_by_model` is now a generic specificity-score loop with no special-casing of `.pt`. (CodeRabbit's own follow-up note on this thread already says: ✅ Addressed in commits d3a57f2 to bd3af6e.)

File: source/tests/pt_expt/infer/test_deep_eval_pt_checkpoint.py [hunk 13]
```diff
@@ -0,0 +1,995 @@
...
+        e, f, v, ae, av = dp.eval(coords, cells, atom_types, atomic=True)
+
+        coord_t = torch.tensor(
+            coords, dtype=torch.float64, device=DEVICE
+        ).requires_grad_(True)
+        atype_t = torch.tensor(
+            atom_types.reshape(1, -1), dtype=torch.int64, device=DEVICE
+        )
+        cell_t = torch.tensor(cells, dtype=torch.float64, device=DEVICE)
+        ref = self.model.forward(coord_t, atype_t, cell_t, do_atomic_virial=True)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/infer/test_deep_eval_pt_checkpoint.py:264]: Valid. Fixed in 9732edbf — `_build_model_and_params` now does `.to(DEVICE)` on the constructed model, mirroring `_make_spin_files`. Tests still pass on CPU; CUDA/MPS runners will no longer device-mismatch in the eager-reference forward path.

**Review by njzjz — APPROVED** [review]
